google/cloud/pubsub/benchmarks/endurance.cc

```diff
@@ -0,0 +1,634 @@
+    sessions.push_back(subscriber.Subscribe(subscription, handler));
+  }
+  auto cleanup_sessions = [&] {
+    for (auto& s : sessions) s.cancel();
+    for (auto& s : sessions) s.wait_for(std::chrono::seconds(3));
+    sessions.clear();
+  };
+
+  auto tasks = [&] {
+    std::vector<std::thread> tasks(config->publisher_count);
+    int task_id = 0;
+    std::generate(tasks.begin(), tasks.end(), [&] {
+      return std::thread{PublisherTask, std::cref(*config),
+                         std::ref(flow_control), task_id++};
+    });
+    return tasks;
+  }();
+
+  auto ts = [] {
+    return google::cloud::internal::FormatRfc3339(
+        std::chrono::system_clock::now());
+  };
+
+  auto divide_duration = [](std::chrono::microseconds duration, int count,
+                            std::chrono::microseconds min,
+                            std::chrono::microseconds max) {
+    auto const d = duration / count;
+    if (d < min) return min;
+    if (d > max) return max;
+    return d;
+  };
+
+  auto const cycle =
+      divide_duration(config->minimum_runtime, 50, std::chrono::seconds(10),
+                      std::chrono::seconds(60));
+
+  auto const report_interval =
+      divide_duration(config->minimum_runtime, 100, std::chrono::seconds(5),
+                      std::chrono::seconds(15));
+
+  auto const start = std::chrono::steady_clock::now();
+  auto report_deadline = start + report_interval;
+
+  std::uniform_int_distribution<std::size_t> session_selector(
+      0, sessions.size() - 1);
+  std::cout << "Timestamp,RunningCount,Count,Min,Max,Average(us)\n";
+
+  Cleanup cleanup;
+  cleanup.Defer(
+      [topic_admin, topic]() mutable { (void)topic_admin.DeleteTopic(topic); });
+  for (auto const& sub : subscriptions) {
+    cleanup.Defer([subscription_admin, sub]() mutable {
+      (void)subscription_admin.DeleteSubscription(sub);
+    });
+  }
+
+  while (!ExperimentCompleted(*config, flow_control, start)) {
+    std::this_thread::sleep_for(cycle);
+    auto const idx = session_selector(generator);
+    auto const& subscription = subscriptions[idx % subscriptions.size()];
+    sessions[idx].cancel();
+    sessions[idx] = subscriber.Subscribe(subscription, handler);
+
+    auto const now = std::chrono::steady_clock::now();
+    if (now >= report_deadline) {
+      report_deadline = now + report_interval;
+      auto const samples = flow_control.ClearSamples();
+      if (samples.empty()) {
+        std::cout << "# " << ts() << ',';
+        flow_control.Debug(std::cout);
+        std::cout << std::endl;
+      } else {
+        auto const p = std::minmax_element(samples.begin(), samples.end());
+        auto const sum = std::accumulate(samples.begin(), samples.end(),
+                                         std::chrono::microseconds{0});
+        auto const mean = [](std::chrono::microseconds sum, std::size_t n) {
+          sum /= n;
+          return sum;
+        }(sum, samples.size());
```

Comment:
Why not simply `auto const mean = sum / samples.size()`?

Reply:
LOL, yes of course.  It used to be more complicated and I failed to simplify it fully.